src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1610 @@
+  // All the links we found in this function. Rarely are there duplicates
+  // in this list (say when writing to the same global location from another
+  // global location), and we do not try to deduplicate here, just store them in
+  // a plain array for now, which is faster (later, when we merge all the info
+  // from the functions, we need to deduplicate anyhow).
+  std::vector<LocationLink> links;
+
+  // All the roots of the graph, that is, places that begin by containing some
+  // particular content. That includes i32.const, ref.func, struct.new, etc. All
+  // possible contents in the rest of the graph flow from such places.
+  //
+  // The vector here is of the location of the root and then its contents.
+  std::vector<std::pair<Location, PossibleContents>> roots;
+
+  // In some cases we need to know the parent of the expression. Consider this:
+  //
+  //  (struct.set $A k
+  //    (local.get $ref)
+  //    (local.get $value)
+  //  )
+  //
+  // Imagine that the first local.get, for $ref, receives a new value. That can
+  // affect where the struct.set sends values: if previously that local.get had
+  // no possible contents, and now it does, then we have DataLocations to
+  // update. Likewise, when the second local.get is updated we must do the same,
+  // but again which DataLocations we update depends on the ref passed to the
+  // struct.get. To handle such things, we set add a childParent link, and then
+  // when we update the child we can find the parent and handle any special
+  // behavior we need there.
+  std::unordered_map<Expression*, Expression*> childParents;
+};
+
+// Walk the wasm and find all the links we need to care about, and the locations
+// and roots related to them. This builds up a CollectedInfo data structure.
+// After all InfoCollectors run, those data structures will be merged and the
+// main flow will begin.
+struct InfoCollector
+  : public PostWalker<InfoCollector, OverriddenVisitor<InfoCollector>> {
+  CollectedInfo& info;
+
+  InfoCollector(CollectedInfo& info) : info(info) {}
+
+  // Check if a type is relevant for us. If not, we can ignore it entirely.
+  bool isRelevant(Type type) {
+    if (type == Type::unreachable || type == Type::none) {
+      return false;
+    }
+    if (type.isTuple()) {
+      for (auto t : type) {
+        if (isRelevant(t)) {
+          return true;
+        }
+      }
+    }
+    if (type.isRef() && getTypeSystem() != TypeSystem::Nominal) {
+      // If nominal typing is not enabled then we cannot handle refs, as we need
+      // to do a subtyping analysis there (which the SubTyping helper class only
+      // supports in nominal mode).
+      return false;
+    }
+    return true;
+  }
+
+  bool isRelevant(Signature sig) {
+    return isRelevant(sig.params) || isRelevant(sig.results);
+  }
+
+  bool isRelevant(Expression* curr) { return curr && isRelevant(curr->type); }
+
+  template<typename T> bool isRelevant(const T& vec) {
+    return true;
```

Comment:
```suggestion
```
Is this a typo?

Reply:
Good catch! Yes, looks like a typo... so the code was considering too many things relevant, and doing wasted work...